File: 27-Interpreter/src/go/or_expression.go
package main

// OrExpression is a Non-terminal Expression — logical OR of two sub-expressions.
type OrExpression struct {
	left  Expression
	right Expression
}

func NewOrExpression(left, right Expression) *OrExpression {
	return &OrExpression{left: left, right: right}
}

func (o *OrExpression) Interpret(context *Context) bool {
	return o.left.Interpret(context) || o.right.Interpret(context)
}

File: 27-Interpreter/src/rust/or_expression.rs
use crate::expression::Expression;
use crate::context::Context;

/// Non-terminal Expression — logical OR of two sub-expressions.
pub struct OrExpression {
    left: Box<dyn Expression>,
    right: Box<dyn Expression>,
}

impl OrExpression {
    pub fn new(left: Box<dyn Expression>, right: Box<dyn Expression>) -> Self {
        OrExpression { left, right }
    }
}

impl Expression for OrExpression {
    fn interpret(&self, context: &Context) -> bool {
        self.left.interpret(context) || self.right.interpret(context)
    }
}
